WordPress/gutenberg
# Sidebar: Avoid focus loss on active tab change

Partially addresses: #8079

This pull request seeks to resolve an issue of focus loss when changing active sidebar tabs for the editor settings sidebar ("Document", "Block").

~Currently, this pull request includes all _but_ the fix itself. Notably, included is an end-to-end test case covering the expected behavior, and a global test event handler protecting against unexpected focus loss.~

**Update:** It has been implemented with https://github.com/WordPress/gutenberg/pull/10917/commits/72569ab71a79dcc777f58a1dffe8db4a9682fef6.

**Implementation notes:**

Focus loss occurs because each specific sidebar renders the entirety of the `.edit-post-wrapper` element. For editor settings, this also includes the tab header as well.

The bug occurs because React is unable to reconcile the two element trees, and is therefore forced to unmount and remount an entirely new sidebar wrapper.

An initial attempt here was to invert rendering, having each specific sidebar populating only the `Fill` of the sidebar component, whereas the `Slot` would be responsible for rendering the `.edit-post-wrapper` wrapper element. This has the added advantage of fixing an issue with the application of `withFocusReturn`, as currently this is broken since it's applied to the `Fill` (non-DOM) element. An issue with this approach is in respecting the `aria-label` application provided by each specific sidebar.

**Testing instructions:**

Verify that no focus loss occurs when changing tabs in the sidebar.

1. Navigate to Posts > Add New
2. Click "Document"
3. Press <kbd>Tab</kbd>
4. Press <kbd>Spacebar</kbd>
5. Note that focus remains on the "Block" tab after activation

Ensure end-to-end tests pass:

```
npm run test-e2e
```
